src/app/OperationalSessionSetup.cpp

```diff
@@ -308,8 +308,21 @@ void OperationalSessionSetup::OnSessionEstablishmentError(CHIP_ERROR error)
     VerifyOrReturn(mState != State::Uninitialized && mState != State::NeedsAddress,
                    ChipLogError(Controller, "HandleCASEConnectionFailure was called while the device was not initialized"));
 
-    DequeueConnectionCallbacks(error);
-    // Do not touch `this` instance anymore; it has been destroyed in DequeueConnectionCallbacks.
+    if (mDiscoveredResults.Allocated())
```

Comment:
Per Slack discussion, let's split up the operational bits and the commissionable discovery bits.